src/main/java/org/ethereum/beacon/discovery/storage/KBucket.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.ethereum.beacon.discovery.storage;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.ethereum.beacon.discovery.schema.NodeRecord;
+import org.ethereum.beacon.discovery.storage.KBuckets.LivenessChecker;
+
+public class KBucket {
+
+  static final int K = 16;
+
+  private final LivenessChecker livenessChecker;
+  private final Clock clock;
+
+  private final List<BucketEntry> nodes = new ArrayList<>();
+
+  private Optional<BucketEntry> pendingNode = Optional.empty();
+
+  public KBucket(final LivenessChecker livenessChecker, final Clock clock) {
+    this.livenessChecker = livenessChecker;
+    this.clock = clock;
+  }
+
+  public List<NodeRecord> getAllNodes() {
+    return nodes.stream().map(BucketEntry::getNode).collect(Collectors.toList());
+  }
+
+  public List<NodeRecord> getLiveNodes() {
+    return nodes.stream()
+        .filter(node -> node.isLive())
+        .map(BucketEntry::getNode)
+        .collect(Collectors.toList());
+  }
+
+  public Optional<NodeRecord> getPendingNode() {
+    return pendingNode.map(BucketEntry::getNode);
+  }
+
+  public void offer(final NodeRecord node) {
+    performMaintenance();
+    getEntry(node)
+        .ifPresentOrElse(
+            existing -> updateExistingRecord(existing, node), () -> offerNewNode(node));
+  }
+
+  private void updateExistingRecord(final BucketEntry existing, final NodeRecord newRecord) {
+    if (existing.getNode().getSeq().compareTo(newRecord.getSeq()) >= 0) {
+      // New record isn't actually newer, do nothing.
+      return;
+    }
+    nodes.remove(existing);
+    final BucketEntry newEntry = new BucketEntry(livenessChecker, newRecord);
+    nodes.add(newEntry);
+    newEntry.checkLiveness(clock.millis());
+  }
+
+  private void offerNewNode(final NodeRecord node) {
+    if (isFull()) {
+      getLastNode().checkLiveness(clock.millis());
+      if (pendingNode.isEmpty()) {
+        livenessChecker.checkLiveness(node);
+      }
+    } else {
+      final BucketEntry newEntry = new BucketEntry(livenessChecker, node);
+      nodes.add(newEntry);
+      newEntry.checkLiveness(clock.millis());
+    }
+  }
+
+  public void onNodeContacted(final NodeRecord node) {
+    getEntry(node)
+        .ifPresentOrElse(
+            existingEntry -> {
+              // Move to the start of the bucket
+              nodes.remove(existingEntry);
+              nodes.add(0, existingEntry.withLastConfirmedTime(clock.millis()));
+              performMaintenance();
+            },
```

Comment:
there's not a LRU cache we can make use of? or just more efficient to maintain our own list?

Reply:
It's not strictly a LRU cache - we need to be quite deliberate about when the node is moved to the front of the queue. We might access and even update the entry for a number of reasons like recording when it was last pinged, which aren't a trigger for moving the node to the front of the bucket as it doesn't confirm the node is live.